src/MergeContext.js

```diff
@@ -855,107 +960,172 @@ class MergeFinalizer extends MergeContext {
+                Config.githubUserEmail(e.email);
+                break;
+            }
+        }
+        assert(Config.githubUserEmail());
+
+        const user = await GH.getUser(Config.githubUserLogin());
+        Config.githubUserName(user.name);
+        assert(Config.githubUserName());
+    }
+
+    async _createStaged() {
+        const baseSha = await GH.getReference(this._prBaseBranchPath());
+        const mergeSha = await GH.getReference("pull/" + this._prNumber() + "/merge");
+        const mergeCommit = await GH.getCommit(mergeSha);
+        if (!Config.githubUserName())
+            await this._acquireUserProperties();
+        let now = new Date();
+        const committer = {name: Config.githubUserName(), email: Config.githubUserEmail(), date: now.toISOString()};
+        const tempCommitSha = await GH.createCommit(mergeCommit.tree.sha, this._prMessage(), [baseSha], mergeCommit.author, committer);
+        this._tagSha = await GH.createReference(tempCommitSha, "refs/" + this._stagingTag());
+        this._compareStatus = "ahead";
+        await this._setApprovalStatus(this._tagSha);
+        await GH.updateReference(Config.stagingBranchPath(), this._tagSha, true);
+    }
+
+    // returns filled StepResult object
+    async _stage() {
+        this._role = "stager";
+
+        this._unlabelPreconditionsChecking();
+        const conditions = await this._checkPreconditions();
+        if (!conditions.succeeded()) {
+            if (conditions.delayed())
+                return conditions;
+            // for failed() or suspended()
+            return StepResult.Fail();
+        }
+        this._unlabelPreconditionsChecked();
+        if (this._dryRun("create staged"))
+            return StepResult.Suspend();
+        await this._createStaged();
+        this._prState = PrState.Staged();
+        this._labelWaitingStagingChecks();
+        return StepResult.Succeed();
+    }
+
     // returns filled StepResult object
-    async process() {
-        const conditions = await this._checkConditions();
-        if (!conditions.succeeded())
-            return conditions;
+    async _mergeStaged() {
+        this._role = "merger";
+
+        let result = await this._checkPostconditions();
+        if (result.succeeded())
+            result = await this._mergeToBase();
+        assert(!result.delayed());
+        return result;
+    }
 
-        const result = await this._mergeToBase();
-        if (!result.succeeded())
-            return result;
+    // Maintain Anubis-controlled PR metadata.
+    // If possible, also merge or advance the PR towards merging.
+    // The caller is responsible for setting PR labels computed here.
+    async _doProcess(anotherPrWasStaged) {
+        await this._loadTag();
+        await this._loadLabels();
+        await this._loadPrState();
+        this._log("PR state calculated: " + this._prState.toString());
+        let result = await this.update();
+        if (anotherPrWasStaged)
```

Comment:
Should we assert that update() never "suspends" the PR (i.e., never prevents another PR from being merged)?